sql/pipelines/src/main/scala/org/apache/spark/sql/pipelines/util/SchemaInferenceUtils.scala

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.pipelines.util
+
+import scala.util.control.NonFatal
+
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.connector.catalog.TableChange
+import org.apache.spark.sql.pipelines.common.DatasetType
+import org.apache.spark.sql.pipelines.graph.{GraphElementTypeUtils, GraphErrors, ResolvedFlow}
+import org.apache.spark.sql.types.{StructField, StructType}
+
+
+object SchemaInferenceUtils {
+
+  /**
+   * Given a set of flows that write to the same destination and possibly a user-specified schema,
+   * we infer the schema of the destination dataset. The logic is as follows:
+   * 1. If there are no incoming flows, return the user-specified schema (if provided)
+   *    or an empty schema.
+   * 2. If there are incoming flows, we merge the schemas of all flows that write to
+   *    the same destination.
+   * 3. If a user-specified schema is provided, we merge it with the inferred schema.
+   *    The user-specified schema will take precedence over the inferred schema.
+   * Returns an error if encountered during schema inference or merging the inferred schema with
+   * the user-specified one.
+   */
+  def inferSchemaFromFlows(
```

Comment:
Let's create a follow-up jira to test all the new utility methods

Reply:
I actually wrote a suite for some of the utilities in there, but it didn't get included in this PR for some reason. I just added it to this branch (`SchemaInferenceUtilsSuite`). I'll create a followup JIRA for the utilities that it doesn't cover.